library/src/main/java/com/github/pwittchen/prefser/library/rx2/Prefser.java

```diff
@@ -280,21 +285,42 @@ public boolean contains(String key) {
    */
   public Observable<String> observePreferences() {
     return Observable.create(new ObservableOnSubscribe<String>() {
+
+      Collection<OnChangeListener> listenerReferences =
+              Collections.synchronizedList(new ArrayList<OnChangeListener>());
+
       @Override
       public void subscribe(final @io.reactivex.annotations.NonNull ObservableEmitter<String> e) {
-        preferences.registerOnSharedPreferenceChangeListener(
-            new SharedPreferences.OnSharedPreferenceChangeListener() {
-              @Override
-              public void onSharedPreferenceChanged(SharedPreferences sharedPrefs, String key) {
-                if (!e.isDisposed()) {
-                  e.onNext(key);
-                }
-              }
-            });
+        final OnChangeListener onChangeListener = new OnChangeListener(e);
+        preferences.registerOnSharedPreferenceChangeListener(onChangeListener);
+        listenerReferences.add(onChangeListener);
+        e.setCancellable(new Cancellable() {
+          @Override
+          public void cancel() {
+            preferences.unregisterOnSharedPreferenceChangeListener(onChangeListener);
+            listenerReferences.remove(onChangeListener);
+          }
+        });
       }
     });
   }
 
+  private static class OnChangeListener
```

Comment:
This is a new class definition, so I'd move it to the end of the file.

Reply:
Sure, will do that - I initially put it here because that's where it was on the Rx1 branch as well so I assumed this to be a project convention.

Reply:
You're right. I'll make it consistent later within another task.